crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1798,6 +1799,61 @@ impl<'db> TypeInferenceBuilder<'db> {
         Type::Unknown
     }
 
+    /// Look up a name reference that isn't bound in the local scope.
+    fn lookup_name(&self, name: &ast::name::Name) -> Type<'db> {
+        let file_scope_id = self.scope.file_scope_id(self.db);
+        let symbols = self.index.symbol_table(file_scope_id);
+        let symbol = symbols
+            .symbol_by_name(name)
+            .expect("Expected the symbol table to create a symbol for every Name node");
+
+        // In function-like scopes, any local variable (symbol that is defined in this
+        // scope) can only have a definition in this scope, or be undefined; it never references
+        // another scope. (At runtime, it would use the `LOAD_FAST` opcode.)
+        if !symbol.is_defined() || !self.scope.is_function_like(self.db) {
+            // Walk up parent scopes looking for a possible enclosing scope that may have a
+            // definition of this name visible to us (would be `LOAD_DEREF` at runtime.)
+            for (enclosing_scope_file_id, _) in self.index.ancestor_scopes(file_scope_id) {
+                // Class scopes are not visible to nested scopes, and we need to handle global
+                // scope differently (because an unbound name there falls back to builtins), so
+                // check only function-like scopes.
+                let enclosing_scope_id = enclosing_scope_file_id.to_scope_id(self.db, self.file);
+                if !enclosing_scope_id.is_function_like(self.db) {
+                    continue;
+                }
```

Comment:
Does this handle e.g.

```pycon
>>> class Foo:
...     x = 1
...     y = [x * 2 for x in range(5)]
... 
>>> Foo.y
[0, 2, 4, 6, 8]
```

?

Reply:
Is that the example you intended to give? I think that example isn't really relevant to this PR; `x` is a defined local inside the comprehension, so there's no nonlocal name reference; the `x = 1` is irrelevant, and this PR shouldn't change anything about the inference.

I think everything in this PR _should_ work correctly for comprehensions; they're just another function-like scope. I'll add some tests to be sure.

Reply:
Argh. You're quite right. Sorry for the noise!